### DISCUSSION TITLE
Is there any value in TestDelegate and ActualValueDelegate?

### DISCUSSION THREAD

**Discussion by Dreamescaper** [nunit/nunit#4823] [General]
Is there any reason not to use `Action` and `Func<T>` instead? Since `Action` and `Func<T>` are natural types for delegates in C#, using other delegates causes some confusion, e.g. here:
```csharp
        var getActual = () => "/* get actual value */";
        Assert.That(getActual, Is.EqualTo("expectedValue").After(10000, 500));
```

Since I'm passing a delegate to Assert method, I'd expect a delegate overload to be invoked. Instead, non-delegate `TActual` overload is used, and assertion fails - because `getActual` is `Func<string>`, not `ActualValueDelegate<TActual>`.

Therefore, my question is. Would it make sense to to replace `TestDelegate` and `ActualValueDelegate` with `Action` and `Func<T>` in the next major version?

**Comment by stevenaw** [comment]
I think this is a great idea, thanks @Dreamescaper 
My guess is that the delegate is there simply as it predates `Func<>` in the BCL, and I think you make a good argument for updating it.
